fix: Modify principal

babyhuey opened the pull request:
I've modified the `__user_email` method to handle more cases where user identity information might be present in the CloudTrail event. The new code will:

- First check if there's a `userIdentity` field at all
- Try to get the email from `principalId` if it exists (common in IAM role cases)
- Try to get the email from `userName` if it exists
- Try to get the email from the `arn` field if it exists
- As a last resort, try to find an email in the entire `userIdentity` object
- Return "Unknown" only if none of the above methods work

This should help capture more user identities and reduce the number of "Unknown" principals in notifications. The code will now handle various ways that AWS might represent user identity in CloudTrail events, including:

- IAM users with email addresses
- IAM roles with assumed role sessions
- Federated users
- AWS service principals

babyhuey committed:
modify principal

babyhuey committed:
Merge branch 'cloudandthings:main' into main

babyhuey committed:
fix: Modify principal

babyhuey committed:
fix: Modify principal

babyhuey committed:
fix: Modify principal

babyhuey commented:
Any chance of this getting looked at anytime soon?

babyhuey commented:
Friendly bump on this - any chance this could get a review? All checks are passing and it's ready to merge. Thanks\!

babyhuey commented:
@baolsen @phzietsman — friendly ping on this one. Would love to get it reviewed when you have a moment. Thanks!

baolsen commented:
Change looks good to me, thanks @babyhuey 
I'll ask Copilot for an additional review but probably good to go :)

baolsen commented:
Alright @babyhuey the AI gods have spoken :)

I agree with the comments for having a check that the value to be returned looks like an email. If it doesn't then the remaining options would be tried. The existing match regex could be re-used for this. 

In terms of tests yea should be simple enough to add some to the existing unit tests.

What do you think - happy to run with it?

babyhuey committed:
fix: validate email format in principalId and userName extraction

Add email regex validation before returning values from principalId
and userName fields. Non-email values (e.g. session names, IAM role
names) now correctly fall through to subsequent extraction strategies
instead of being returned as the user email.

Add unit tests covering email and non-email cases for all extraction
paths.

babyhuey committed:
fix: add logging to email extraction and keep non-email userName as fallback

Add debug logging at each decision point in the email extraction
pipeline so operators can diagnose why a particular identity was
returned.

Keep non-email userName as a last-resort fallback before returning
"Unknown", so notifications still show a useful identifier (e.g.
role name) when no email can be found.

babyhuey commented on clickopsnotifier/clickops.py:
```diff
@@ -16,17 +16,35 @@ def __init__(self, event) -> None:
+
+        # Try to get email from principalId if it exists
+        if "principalId" in user_identity:
+            # Handle cases like "AROAXK4KVD27BINQTHSKU:paul@cloudandthings.io"
+            parts = user_identity["principalId"].split(":")
+            if len(parts) > 1:
+                return parts[1]
```
Done — principalId now validates the part after the colon against the email regex using `fullmatch()` before returning it. Non-email session names fall through to the next extraction strategy.

babyhuey commented on clickopsnotifier/clickops.py:
```diff
@@ -16,17 +16,35 @@ def __init__(self, event) -> None:
+
+        # Try to get email from userName if it exists
+        if "userName" in user_identity:
+            return user_identity["userName"]
```
Done — userName is now validated against the email regex first. If it's not an email, extraction continues through arn and the full userIdentity scan. As a last resort, the non-email userName is still returned as a fallback (preferable to "Unknown" for operator triage).

babyhuey commented on clickopsnotifier/clickops.py:
```diff
@@ -16,17 +16,35 @@ def __init__(self, event) -> None:
+
+        # Try to get email from userName if it exists
+        if "userName" in user_identity:
+            return user_identity["userName"]
```
Done — added 12 unit tests covering: non-email principalId (session name), non-email userName (role name), cascading fallthrough from principalId to userName to arn, and the non-email userName fallback when no email is found anywhere.

babyhuey committed:
fix: apply black formatting and update deployment packages

Reformat clickops.py and test_user_email.py to comply with
black 22.6.0. Update all deployment zip packages to include
the latest clickops.py.

babyhuey commented:
Fixed CI failures:

- Reformatted code to comply with black 22.6.0
- Updated clickops.py in all deployment zip packages to match source

babyhuey committed:
fix: flake8 line length violations and update README terraform docs

Fix E501 line too long errors in test_user_email.py by extracting
long ARN strings into variables. Update README.md lambda_runtime
docs to match current variable defaults.

babyhuey commented:
Sorry about the CI churn! Should be sorted now. Here's what I verified locally in a Python 3.11.2 Docker container matching CI:

- **flake8 5.0.4**: Zero violations on all changed files (the two E501 line-too-long errors in test_user_email.py are fixed)
- **black 22.6.0**: All changed files pass `--check` with no reformatting needed
- **Deployment zips**: All 5 deployment packages updated and matching source
- **README.md**: Updated terraform-docs for `lambda_runtime` to match current variable defaults (`python3.11`)

The remaining `terraform_validate` failures in pre-commit (`examples/delivery_stream` and `examples/standalone`) are pre-existing — they're caused by conflicting AWS provider constraints (`~> 4.9` vs `>= 6.0.0`) in the example modules' dependencies, unrelated to this PR.

baolsen commented:
No worries @babyhuey, I'll log a separate issue to fixup the CI as it needs some housekeeping.

Thanks for pointing that out and apologies for the runaround.

I'm going to merge and release this one so you can test the new version on your side.

babyhuey commented:
FYI the only remaining pre-commit failure is `terraform_validate` on `examples/delivery_stream` and `examples/standalone`. This is pre-existing and unrelated to this PR.

The issue is `trussworks/logs/aws v18.0.0` now requires `aws >= 6.0.0`, which conflicts with `~> 4.9` in both examples. Quick fix would be to change the AWS provider constraint in those two example `main.tf` files from:

```hcl
version = "~> 4.9"
```

to:

```hcl
version = "~> 6.0"
```

Happy to open a separate PR for that if you'd like.

baolsen commented:
> FYI the only remaining pre-commit failure is `terraform_validate` on `examples/delivery_stream` and `examples/standalone`. This is pre-existing and unrelated to this PR.
> 
> The issue is `trussworks/logs/aws v18.0.0` now requires `aws >= 6.0.0`, which conflicts with `~> 4.9` in both examples. Quick fix would be to change the AWS provider constraint in those two example `main.tf` files from:
> 
> ```hcl
> version = "~> 4.9"
> ```
> 
> to:
> 
> ```hcl
> version = "~> 6.0"
> ```
> 
> Happy to open a separate PR for that if you'd like.

No stress, I'll get it sorted as part of #109
